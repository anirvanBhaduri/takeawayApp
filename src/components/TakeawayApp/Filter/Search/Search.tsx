import * as React from 'react';

import {Filter, FilterProps} from "../Filter";
import {RestaurantProps} from "../../RestaurantList/Restaurant/Restaurant";

/**
 * The searchTerm is required to ensure
 * we can search.
 *
 * @interface SearchProps
 */
interface SearchProps extends FilterProps {
    searchTerm: string;
}

/**
 * Search using a string.
 *
 * Used to filter RestaurantProps.
 *
 * @class Search
 * @package TakeawayApp
 *
 * @author Anirvan Bhaduri
 * @since 9th Dec 2018
 * @version 0.0.1
 */
export class Search extends Filter<SearchProps> {

    /**
     * The search term used to filter the results.
     */
    searchTerm: string;

    /**
     * Search constructor.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        this.searchTerm = this.props.searchTerm;
    }

    /**
     * Set the search term.
     *
     * @param searchTerm
     */
    setSearchTerm(searchTerm: string) {
        this.searchTerm = searchTerm;
    }

    /**
     * Get the current search term.
     *
     * @return string
     */
    getSearchTerm(): string {
        return this.searchTerm;
    }

    /**
     * Apply the search filter.
     *
     * @param list
     *
     * @return RestaurantProps
     */
    applyFilter(list: RestaurantProps[]): RestaurantProps[] {
        // Do some filter logic here using searchTerm
        // ... should mainly remove elements

        // When we don't have a search term, we don't want to filter at all
        if (!this.searchTerm) {
            return list;
        }

        const searchTerm = this.searchTerm.toLowerCase();
        return list.filter(
            (value: RestaurantProps): boolean => {
                return value.name.toLowerCase().indexOf(searchTerm) >= 0;
            }
        );
    }

    /**
     * Render the search component.
     *
     * @return JSX
     */
    render() {
        const handler = this.props.filterHandler;
        const searchTerm = this.props.searchTerm;

        return (
            <div>
                <input
                    onChange={handler}
                    value={searchTerm}
                    type="input"
                />
            </div>
        );
    }
}
