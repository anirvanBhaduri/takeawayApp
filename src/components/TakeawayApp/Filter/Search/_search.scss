/// Search component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import '../../../UI/_colour_constants.scss';
@import '../../../UI/size_constants';

// The search bar's styles
.search {
  float: left;
  width: 550px;

  @media screen and (max-width: $screen-break-medium) {
    width: 98%;
    margin: 10px;
  }

  @media screen and (max-width: $screen-break-small) {
    width: 98%;
    margin: 10px;
  }
}

// The search bar's input's styles
.search__input {
  display: inline-block;
  @include input-styles(28px, 545px);

  @media screen and (max-width: $screen-break-medium) {
    font-size: inherit;
  }

  @media screen and (max-width: $screen-break-small) {
    font-size: inherit;
  }
}