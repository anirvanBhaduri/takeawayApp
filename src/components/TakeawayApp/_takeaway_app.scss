/// TakeAwayApp component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import 'UI/colours';
@import 'UI/fonts';
@import 'UI/dimensions';

// Render main container
.takeaway-app {
  background-color: $theme-background;
  height: $app-height;
  padding: 10px;
  font-family: $container-font-family;
  border: 2px solid $takeaway-color;
  border-radius: 16px;
  text-align: center;

  @media screen and (min-width: $screen-break-width) {
    padding-left: 100px;
    padding-right: 100px;
  }
}

// Title element for takeaway component
.takeaway-app__title {
  font-style: italic;
}

// Body element for takeaway component
.takeaway-app__body {
  width: $app-body-width;
  height: 90%;
  margin: 0;
  display: inline-block;
  text-align: center;
}
