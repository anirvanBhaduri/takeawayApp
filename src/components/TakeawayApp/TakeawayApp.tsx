import * as React from "react";

import {RestaurantList} from "./RestaurantList/RestaurantList";
import {Search} from "./Filter/Search/Search";
import {Sort} from "./Filter/Sort/Sort";
import {sortables} from "./Filter/Sort/Sortables";
import {OpeningState} from "./Filter/OpeningState/OpeningState";
import {Favourite} from "./Filter/Favourite/Favourite";

// The sample data.
const sample = require('../../../dist/Sample.json');

/**
 * Define the format for the TakeawayApp state.
 *
 * @interface TakeawayAppState
 */
interface TakeawayAppState {
    restaurants: [];
    filterCriteria: {[index: string]: any};
}

/**
 * The Takeaway app.
 *
 * This is the top level container. For the purpose
 * of this assignment, we will not use Redux due to the app's simplicity.
 * However, as this app grows, it will likely become necessary
 * to move the logic to redux actions, reducers etc.
 *
 * @class TakeawayApp
 * @package TakeawayApp
 *
 * @author Anirvan Bhaduri
 * @since 9th Dec 2018
 * @version 0.0.1
 */
export class TakeawayApp extends React.Component {

    /**
     * A list of restaurants
     *
     * @property [] restaurants
     */
    restaurants: [] = [];

    /**
     * Initialise app state.
     *
     * @property TakeawayAppState state
     *
     * - searchCriteria - a list of search criteria defining what to display
     *                      and the order to display them
     *                  - the list sorts/filters top down, where top is of least importance and
     *                      bottom the most
     * - restaurants - a list of restaurants. This is expected to stay constant through
     *                  this app.
     */
    state: TakeawayAppState = {
        filterCriteria: {
            search: new Search({
                searchTerm: '',
                filterHandler: () => {}
            }),
            sort: new Sort({
                sortType: Object.keys(sortables)[0],
                filterHandler: () => {}
            }),
            openingState: new OpeningState({
                filterHandler: () => {}
            }),
            favourite: new Favourite({
                filterHandler: () => {}
            }),
        },
        restaurants: [],
    };

    /**
     * Construct TakeawayApp.
     *
     * Lets load in the restaurants here.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        // Load sample restaurants.
        const {restaurants} = sample;
        this.state.restaurants = restaurants;

        // Bind handlers.
        this.handleSearchChange = this.handleSearchChange.bind(this);
        this.handleSortChange = this.handleSortChange.bind(this);
    }

    /**
     * Using the event, change the search filterCriteria
     * to use the updated search value.
     *
     * Empty means no search.
     *
     * @param e
     */
    handleSearchChange(e: any) {
        const {filterCriteria} = this.state;

        filterCriteria.search.setSearchTerm(e.target.value);

        this.setState({filterCriteria: filterCriteria});
    }

    /**
     * Using the event, change the sort filterCriteria
     * to use the updated sort value.
     *
     * @param e
     */
    handleSortChange(e: any) {
        // Lets handle the sort selection here.
        // ...
        const sortValue = e.target.value;
        const {filterCriteria} = this.state;

        filterCriteria.sort.setSortType(sortValue);
        this.setState({filterCriteria: filterCriteria});
    }

    /**
     * Render the TakeawayApp.
     *
     * @return JSX
     */
    render() {
        const filterCriteria = this.state.filterCriteria;
        const restaurants = this.state.restaurants;
        const search = this.state.filterCriteria.search;

        return (
            <div className={'takeaway-app'} >
                <h2 className={'takeaway-app__title'} >Find a Restaurant</h2>
                <div className={'takeaway-app__body'} >
                    <Search searchTerm={search.getSearchTerm()} filterHandler={this.handleSearchChange} />
                    <Sort sortType={''} filterHandler={this.handleSortChange} />
                    <RestaurantList restaurants={restaurants} filterCriteria={filterCriteria} />
                </div>
            </div>
        )
    }
}
