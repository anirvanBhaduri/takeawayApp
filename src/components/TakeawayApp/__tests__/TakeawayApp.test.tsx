import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

// SUT
import {TakeawayApp} from '../TakeawayApp';
import {Search} from "../Filter/Search/Search";

/**
 * TakeawayApp component tests.
 *
 * @covers TakeawayApp
 */
describe('Takeaway Component Tests', () => {

    /**
     * TakeawayApp should render the correct components.
     *
     * @covers TakeawayApp.render
     */
    it('should render a Search, Sort and Restaurant List', () => {
        const wrapper = shallow(<TakeawayApp />);

        expect(wrapper.find('Search').exists()).toBe(true);
    });

    /**
     * TakeawayApp renders correctly.
     *
     * @covers TakeawayApp.render
     */
    it('should render correct snapshot', () => {
        const wrapper = shallow(<TakeawayApp />);

        expect(wrapper).toMatchSnapshot();
    })
});