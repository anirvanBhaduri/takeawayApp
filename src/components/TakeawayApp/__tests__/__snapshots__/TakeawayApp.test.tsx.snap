// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Takeaway Component Tests should render correct snapshot 1`] = `
<div
  className="takeaway-app"
>
  <h2
    className="takeaway-app__title"
  >
    Find a Restaurant
  </h2>
  <div
    className="takeaway-app__body"
  >
    <Search
      filterHandler={[Function]}
      searchTerm=""
    />
    <Sort
      filterHandler={[Function]}
      sortType=""
    />
    <RestaurantList
      filterCriteria={
        Object {
          "favourite": Favourite {
            "context": undefined,
            "props": Object {
              "filterHandler": [Function],
            },
            "refs": Object {},
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
          "openingState": OpeningState {
            "context": undefined,
            "props": Object {
              "filterHandler": [Function],
            },
            "refs": Object {},
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
          "search": Search {
            "context": undefined,
            "props": Object {
              "filterHandler": [Function],
              "searchTerm": "",
            },
            "refs": Object {},
            "searchTerm": "",
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
          "sort": Sort {
            "context": undefined,
            "props": Object {
              "filterHandler": [Function],
              "sortType": "bestMatch",
            },
            "refs": Object {},
            "sortType": "bestMatch",
            "sortables": Array [
              <option
                className="sort__option"
                value="bestMatch"
              >
                Best Match
              </option>,
              <option
                className="sort__option"
                value="newest"
              >
                Newest
              </option>,
              <option
                className="sort__option"
                value="ratingAverage"
              >
                Rating Average
              </option>,
              <option
                className="sort__option"
                value="distance"
              >
                Distance
              </option>,
              <option
                className="sort__option"
                value="popularity"
              >
                Popularity
              </option>,
              <option
                className="sort__option"
                value="averageProductPrice"
              >
                Average Product Price
              </option>,
              <option
                className="sort__option"
                value="deliveryCosts"
              >
                Delivery Costs
              </option>,
              <option
                className="sort__option"
                value="minCost"
              >
                Min Cost
              </option>,
              <option
                className="sort__option"
                value="topRestaurant"
              >
                Top Restaurant
              </option>,
            ],
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
        }
      }
      restaurants={
        Array [
          Object {
            "name": "Tanoshii Sushi",
            "sortingValues": Object {
              "averageProductPrice": 1536,
              "bestMatch": 0,
              "deliveryCosts": 200,
              "distance": 1190,
              "minCost": 1000,
              "newest": 96,
              "popularity": 17,
              "ratingAverage": 4.5,
            },
            "status": "open",
          },
          Object {
            "name": "Tandoori Express",
            "sortingValues": Object {
              "averageProductPrice": 1146,
              "bestMatch": 1,
              "deliveryCosts": 150,
              "distance": 2308,
              "minCost": 1300,
              "newest": 266,
              "popularity": 123,
              "ratingAverage": 4.5,
            },
            "status": "closed",
          },
          Object {
            "name": "Royal Thai",
            "sortingValues": Object {
              "averageProductPrice": 1492,
              "bestMatch": 2,
              "deliveryCosts": 150,
              "distance": 2639,
              "minCost": 2500,
              "newest": 133,
              "popularity": 44,
              "ratingAverage": 4.5,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Sushi One",
            "sortingValues": Object {
              "averageProductPrice": 1285,
              "bestMatch": 3,
              "deliveryCosts": 0,
              "distance": 1618,
              "minCost": 1200,
              "newest": 238,
              "popularity": 23,
              "ratingAverage": 4,
            },
            "status": "open",
          },
          Object {
            "name": "Roti Shop",
            "sortingValues": Object {
              "averageProductPrice": 915,
              "bestMatch": 4,
              "deliveryCosts": 0,
              "distance": 2308,
              "minCost": 2000,
              "newest": 247,
              "popularity": 81,
              "ratingAverage": 4.5,
            },
            "status": "open",
          },
          Object {
            "name": "Aarti 2",
            "sortingValues": Object {
              "averageProductPrice": 922,
              "bestMatch": 5,
              "deliveryCosts": 250,
              "distance": 1605,
              "minCost": 500,
              "newest": 153,
              "popularity": 44,
              "ratingAverage": 4.5,
            },
            "status": "open",
          },
          Object {
            "name": "Pizza Heart",
            "sortingValues": Object {
              "averageProductPrice": 1103,
              "bestMatch": 6,
              "deliveryCosts": 150,
              "distance": 2453,
              "minCost": 1500,
              "newest": 118,
              "popularity": 9,
              "ratingAverage": 4,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Mama Mia",
            "sortingValues": Object {
              "averageProductPrice": 912,
              "bestMatch": 7,
              "deliveryCosts": 0,
              "distance": 1396,
              "minCost": 1000,
              "newest": 250,
              "popularity": 6,
              "ratingAverage": 4,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Feelfood",
            "sortingValues": Object {
              "averageProductPrice": 902,
              "bestMatch": 8,
              "deliveryCosts": 150,
              "distance": 2732,
              "minCost": 1500,
              "newest": 163,
              "popularity": 31,
              "ratingAverage": 4.5,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Daily Sushi",
            "sortingValues": Object {
              "averageProductPrice": 1327,
              "bestMatch": 9,
              "deliveryCosts": 200,
              "distance": 1911,
              "minCost": 1000,
              "newest": 221,
              "popularity": 6,
              "ratingAverage": 4,
            },
            "status": "closed",
          },
          Object {
            "name": "Pamukkale",
            "sortingValues": Object {
              "averageProductPrice": 968,
              "bestMatch": 10,
              "deliveryCosts": 0,
              "distance": 2353,
              "minCost": 2000,
              "newest": 201,
              "popularity": 25,
              "ratingAverage": 4,
            },
            "status": "closed",
          },
          Object {
            "name": "Indian Kitchen",
            "sortingValues": Object {
              "averageProductPrice": 1189,
              "bestMatch": 11,
              "deliveryCosts": 150,
              "distance": 2308,
              "minCost": 1300,
              "newest": 272,
              "popularity": 5,
              "ratingAverage": 4.5,
            },
            "status": "open",
          },
          Object {
            "name": "CIRO 1939",
            "sortingValues": Object {
              "averageProductPrice": 1762,
              "bestMatch": 12,
              "deliveryCosts": 99,
              "distance": 3957,
              "minCost": 1300,
              "newest": 231,
              "popularity": 79,
              "ratingAverage": 4.5,
            },
            "status": "open",
          },
          Object {
            "name": "Zenzai Sushi",
            "sortingValues": Object {
              "averageProductPrice": 1579,
              "bestMatch": 13,
              "deliveryCosts": 0,
              "distance": 2911,
              "minCost": 2000,
              "newest": 155,
              "popularity": 36,
              "ratingAverage": 4,
            },
            "status": "closed",
          },
          Object {
            "name": "Fes Patisserie",
            "sortingValues": Object {
              "averageProductPrice": 1214,
              "bestMatch": 14,
              "deliveryCosts": 150,
              "distance": 2302,
              "minCost": 1250,
              "newest": 77,
              "popularity": 3,
              "ratingAverage": 4,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Yvonne's Vispaleis",
            "sortingValues": Object {
              "averageProductPrice": 2557,
              "bestMatch": 15,
              "deliveryCosts": 150,
              "distance": 2909,
              "minCost": 1750,
              "newest": 150,
              "popularity": 3,
              "ratingAverage": 5,
            },
            "status": "order ahead",
          },
          Object {
            "name": "De Amsterdamsche Tram",
            "sortingValues": Object {
              "averageProductPrice": 892,
              "bestMatch": 304,
              "deliveryCosts": 0,
              "distance": 2792,
              "minCost": 0,
              "newest": 131,
              "popularity": 0,
              "ratingAverage": 0,
            },
            "status": "open",
          },
          Object {
            "name": "Lale Restaurant & Snackbar",
            "sortingValues": Object {
              "averageProductPrice": 838,
              "bestMatch": 305,
              "deliveryCosts": 0,
              "distance": 2880,
              "minCost": 0,
              "newest": 73,
              "popularity": 0,
              "ratingAverage": 0,
            },
            "status": "order ahead",
          },
          Object {
            "name": "Lunchpakketdienst",
            "sortingValues": Object {
              "averageProductPrice": 4465,
              "bestMatch": 306,
              "deliveryCosts": 500,
              "distance": 14201,
              "minCost": 5000,
              "newest": 259,
              "popularity": 0,
              "ratingAverage": 3.5,
            },
            "status": "open",
          },
        ]
      }
    />
  </div>
</div>
`;
